src/analysis/processing/qgsalgorithmrandomextract.cpp

```diff
@@ -0,0 +1,136 @@
+{
+  return new QgsRandomExtractAlgorithm();
+}
+
+void QgsRandomExtractAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ),
+                QList< int >() << QgsProcessing::TypeVector ) );
+  addParameter( new QgsProcessingParameterEnum( QStringLiteral( "METHOD" ), QObject::tr( "Method" ), QStringList() << QObject::tr( "Number of features" ) << QObject::tr( "Percentage of features" ), false, 0 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "NUMBER" ), QObject::tr( "Number/percentage of features" ),
+                QgsProcessingParameterNumber::Integer, 10, false, 0 ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Extracted (random)" ) ) );
+}
+
+QVariantMap QgsRandomExtractAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, source->fields(),
+                                          source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  int method = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+  int number = parameterAsInt( parameters, QStringLiteral( "NUMBER" ), context );
+
+  long count = source->featureCount();
+
+  if ( method == 0 )
+  {
+    // number of features
+    if ( number > count )
+      throw QgsProcessingException( QObject::tr( "Selected number is greater than feature count. Choose a lower value and try again." ) );
+  }
+  else
+  {
+    // percentage of features
+    if ( number > 100 )
+      throw QgsProcessingException( QObject::tr( "Percentage can't be greater than 100. Choose a lower value and try again." ) );
+
+    number = static_cast< int >( std::ceil( number * count / 100 ) );
+  }
+
+  // initialize random engine
+  std::random_device randomDevice;
+  std::mt19937 mersenneTwister( randomDevice() );
+  std::uniform_int_distribution<int> fidsDistribution( 0, count );
+
+  QVector< QgsFeatureId > fids( number );
+  std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
+
+  double step = number > 0 ? 100.0 / number : 1;
+  int i = 0;
+
+  QgsFeature f;
+  for ( QgsFeatureId id : fids )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    source->getFeatures( QgsFeatureRequest( id ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks ).nextFeature( f );
```

Comment:
One by one feature requests are always quite slow. Looks like the original alg looped through the whole layer to avoid this. So while the new code will be much faster for low selection counts, when the count get close to the feature count it'll be much slower. 

I think you could avoid this by creating a single feature request with a .setFilterFids on it, and iterating through that. Note that you'd need to count the number of times that feature appears in the fids list and add that many duplicate features though (or you'll reintroduce the original bug!)

On that note, I think it might be worthwhile exposing a enum choice for sampling behaviour -- i.e. "sample with replacement" or "sample without replacement".

Reply:
My initial idea was exactly the same - use fids filter to query only required features. Unfortunately this way we loose duplicate features if no additional logic will be added. So I decided to go for slower option for now. Will look at it again and see if I can make it better.

Reply:
Out of curiosity, would there be a way to do one request for the base set and another round for duplicates by combining the fids with an expression?

Reply:
@roya0045 have you read comments above? This will be addressed in the follow-up pull-request

Reply:
@alexbruy yes but I was just trying to suggest a way to do things eventually.

Reply:
this was already suggested above

> you could avoid this by creating a single feature request with a .setFilterFids on it, and iterating through that. Note that you'd need to count the number of times that feature appears in the fids list and add that many duplicate features though (or you'll reintroduce the original bug!)

Reply:
This was not quite what I had in mind, I was talking about a first phase using the fids, and then for the second pass use fids + an expression to obtain the duplicates that you need instead of doing the exact same query twice or multiple times.

By having those two lists you would not have to care about reading things multiple times.

But this hinges on the fact that there is an optimal way to do this by using an expression and I am not familiar enough with the alg to know if this is true, hence why I was asking.